### PR TITLE
Update the contributing guidelines

### PR DESCRIPTION
**Pull request by floriank**
## Goal

In the light of some of the latest pull requests it became apparent hat we lack a single source of truth when it comes to the guidelines for contributing to OpenProject.

After some reviewing and looking into the history, I updated the guidelines for contributing to OpenProject to reflect current practices.
## Notes

Information has been pulled together from different source, mainly the [Guidelines from the OpenProject](https://community.openproject.org/projects/openproject/wiki/Git_Workflow) wiki (which is no longer available to me), comments made by @linki in #1362 and the discussion in #2594, as well as discussion with @ulferts on current practices in handling PRs.

There also was a discussion with @myabc last week on using [CLAHub](https://www.clahub.com/) for contributors to the repository, to which these guidelines have no reflection of. I'd be happy on discussing this topic, as there is some need of that in my opinion.
## Ticket

https://community.openproject.org/work_packages/18827

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://floriank-fix-update-contributing.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-fix-update-contributing.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,7 +1,50 @@
 OpenProject is an open source project and we encourage you to help us out. We'd be happy if you do one of these things:
```
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: I think you should also reword this intro or keep the list of things that someone might want to do to help us (create tickets, be active in forum, etc.).

**tl; dr:** Your write-up below does not match this paragraph

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,7 +1,50 @@
...
+- Clone your fork to your development machine: 
+
+```
+git clone git@github.com/<username>/openproject
+```
+
+- Optional: Add the original OpenProject repository as a remote, so you can fetch changes: 
+
+```
+git remote add upstream git@github.com:/opf/openproject
```
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: I'd say the `:` after `github.com` is superfluous.

Comment (floriank) [inline CONTRIBUTING.md]: I'd say I just copied it from the now unavailable git_Workflow from the OpenProject wiki page :laughing:

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,7 +1,50 @@
...
+```
+
+- Make sure you're on the right branch, right now (February 2015), the main development branch is dev:
```
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: I don't think that the timestamp (February 2015) should be neccessary. This information is not allowed to out-date. We will only get uncertainty by such phrases... Better:

> Make sure you're on the right branch. The main development branch is dev

Comment (NobodysNightmare) [inline CONTRIBUTING.md]: Also: What if I have something from what I believe should be included in a backport branch?

Is this possible at all? What should I do, which limitations are there? How is this backport branch called?

File: CONTRIBUTING.md (hunk 1 above)
Comment (floriank) [inline CONTRIBUTING.md]: Funnily enough, I had a different version which got lost in my git-fu.

I'd reword it to something along the lines of:

"If you want to help us, please read the following guidelines to contributing:"

Optionally we could reword the individual headlines to be more "pro-active", like "Creating tickets in OP", "Creating new features", "Creating hotfixes", etc.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,7 +1,50 @@
...
+
+- Create a pull request (PR) against a branch of of the <opf/openproject> repository, containing a ***clear description*** of what the pull request attempts to change and/or fix. 
+
+We will then review your PR. Please note that you can add commits after the PR has been created by pushing to the branch in your fork.
+
+## Important notices
+
+- Please add tests to your code to verify functionality, especially if it is a new feature. Please also run these tests locally.
+- Please create pull requests against the current `dev` branch. Hotfixes and Bugfixes should be created against the appropiate `release/*` branch.
+- We want to keep the Pull request list as cleaned up as possible - we will close pull requests after an inactivity period of 48 hours (no comments, no further pushes)
```
Comment (Azure7111) [inline CONTRIBUTING.md]: Although I like the idea of closing inactive pull requests we might have to work out something more flexible here in terms of when to close a PR. Think about the holidays people usually enjoy at the end of a year, or a sudden increase in contributions. Surely there might be even more things to concern when designing something around inactivity. On the other hand this might lead to very complex rules that are difficult to understand, therefore if it works we could probably just handle this rule less strictly.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,7 +1,50 @@
...
-* Help out other people on our [forums](https://community.openproject.org/projects/openproject/boards)
-* Help us [translate OpenProject to more languages](https://community.openproject.org/projects/openproject/wiki/Translations)
-* Contribute code via GitHub Pull Requests, see our [contribution page](https://community.openproject.org/projects/openproject/wiki/Contribution) for more information
-* See [Git Flow](https://community.openproject.org/projects/openproject/wiki/Git_Branching_Model)
+## Issue flow
+
+Ideally, there is a work package for every issue that is tackled by a contributor - the list can be found [here](https://community.openproject.org/projects/openproject/work_packages). Work packages can be created there on-demand if necessary.
+
+## Development flow
+For contributing source code, please follow the Git Workflow below:
```
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: We should probably add some wording to indicate, that the rules below apply to community contributions.

While the OpenProject core developers should also try to stick to these rules, they might deviate from them, e.g. not fork the repo, but have a branch on the main repository.

Proposal:

> For external contributions [...] please follow the Git Workflow below:

File: CONTRIBUTING.md (hunk 4 above)
Comment (floriank) [inline CONTRIBUTING.md]: I added this speciffically to avoid aging PRs.

My suggestion would to rephrase it to something along the lines of "The aim is to clouse PRs that are inactive for more than 72 hours which do not have a WIP label".

Ideally, the goal should be not to have multiple active working sites concurrently. Overall it should reduce the number of PRs, especially inactive ones.

File: CONTRIBUTING.md (hunk 2 above)
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: Better copy from GitHub:

```
https://github.com/opf/openproject.git
```

note that I obtained that link from a private tab (i.e. not being logged in, not having write-access to the repo). I think SSH URL only works with write access, but I am not sure

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,7 +1,46 @@
...
+- Create a pull request (PR) against a branch of of the <opf/openproject> repository, containing a ***clear description*** of what the pull request attempts to change and/or fix. 
+
+We will then review your PR. Please note that you can add commits after the PR has been created by pushing to the branch in your fork.
+
+## Important notices
+
+- Please add tests to your code to verify functionality, especially if it is a new feature. Please also run these tests locally.
+- Please create pull requests against the current `dev` branch. Hotfixes and Bugfixes should be created against the appropiate `release/*` branch.
+- We want to keep the Pull request list as cleaned up as possible - we will close pull requests after an inactivity period of 48 hours (no comments, no further pushes)
+- We use OpenProject for development coordination - please have a look at [the work packages list](https://community.openproject.org/projects/openproject/work_packages) for upcoming features and reported bugs.
```
Comment (Azure7111) [inline CONTRIBUTING.md]: Just noticed that the linked WP list also contains test tickets. Maybe it should include a small filter for that to avoid confusion.

File: CONTRIBUTING.md (hunk 4 above)
Comment (Azure7111) [inline CONTRIBUTING.md]: That might just lead to stale PR's having a WIP label tacked on.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,7 +1,46 @@
...
 
-* Create a [new work package on openproject.org](https://community.openproject.org/projects/openproject/work_packages) if you find a bug or need a feature
-* Help out other people on our [forums](https://community.openproject.org/projects/openproject/boards)
-* Help us [translate OpenProject to more languages](https://community.openproject.org/projects/openproject/wiki/Translations)
-* Contribute code via GitHub Pull Requests, see our [contribution page](https://community.openproject.org/projects/openproject/wiki/Contribution) for more information
-* See [Git Flow](https://community.openproject.org/projects/openproject/wiki/Git_Branching_Model)
+## Development flow
+For contributing source code, please follow the Git Workflow below:
+
+- Fork Openproject on GitHub
```
Comment (myabc) [inline CONTRIBUTING.md]: Open**P**roject

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,7 +1,46 @@
...
+```
+
+- Create a pull request (PR) against a branch of of the <opf/openproject> repository, containing a ***clear description*** of what the pull request attempts to change and/or fix. 
+
+We will then review your PR. Please note that you can add commits after the PR has been created by pushing to the branch in your fork.
+
+## Important notices
+
+- Please add tests to your code to verify functionality, especially if it is a new feature. Please also run these tests locally.
+- Please create pull requests against the current `dev` branch. Hotfixes and Bugfixes should be created against the appropiate `release/*` branch.
```
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: I think the information around non-dev PRs is incomplete:
- How do I find out which branch is the right one?
- Which kinds of PRs are allowed there (e.g. no new features)
- Should I make two PRs in such a case? (one against `release/x` and one against `dev`?)
  - the answer is no... but I can't see it ^^

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,7 +1,46 @@
...
+- We want to keep the Pull request list as cleaned up as possible - we will aim close pull requests after an **inactivity period of 72 hours** (no comments, no further pushes) which are not labelled as WIP.
```
Comment (Azure7111) [inline CONTRIBUTING.md]: Suggestion: Change `WIP` to `work in progress` as this matches the exact wording of the corresponding label.

Comment (NobodysNightmare) [inline CONTRIBUTING.md]: Also note that only contributors (**we**) are able to assign labels.

**Comment by NobodysNightmare** [comment]
We require a CLA to be signed, which you do not currently mention anywhere.
I'll ask Birthe whether [CLAHub](https://www.clahub.com/) is an option for us (especially legally). Otherwise the correct way should be something like sending a letter o.O

edit: One thing that I find worrying about CLAHub is that the last activity there was two years ago...

File: CONTRIBUTING.md (hunk 8 above)
Comment (floriank) [inline CONTRIBUTING.md]: You have to help me out here. My understanding after a couple of weeks:
1. Features/gem updates go against `dev`
2. Hotfixes go against latest `release/*`
3. Bugfixes go into `dev`
4. Backports (bugfixes/hotfixes) go into `release/*`

**Comment by myabc** [comment]
> Otherwise the correct way should be something like sending a letter o.O

Electronically-signed CLA's are used by a lot of big companies now. IANAL, but I'm pretty sure electronic signatures are also valid in our jurisdiction.

If we're required to have paper CLA's – sent my post – we might as well just write _f**\* you. we don't take contributions_ :stuck_out_tongue_winking_eye:

**Comment by NobodysNightmare** [comment]
We are in germany here... I am not sure what qualifies as electronic signature either ^^

And what does IANAL mean?

**Comment by NobodysNightmare** [comment]
After actually **reading** the [CLA](https://www.openproject.org/wp-content/uploads/2014/09/OPF-Contributor-License-Agreement_v.2.pdf) I realized, that the CLA states how it should be transferred:

> Please complete the agreement and send it together with a detached gpg signature via email to cla@openproject.org

This should be the way then... until we decide internally, that CLAHub is also fine...

**Comment by floriank** [comment]
I am not a lawyer. IANAL.

**Comment by NobodysNightmare** [comment]
What about Rubocop/Hound? Do I have to care about code style and coding conventions?

**Comment by floriank** [comment]
I would add a section for style on this for new contributions, i.e. lines you touch. 
This would include a note on hound as well as using `rubocop -a` on the files to be commited.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,7 +1,102 @@
...
+
+### Code style guidelines
+
+We use [RuboCop](https://github.com/bbatsov/rubocop) to verify our code style for Ruby. A [RuboCop configuration](https://github.com/opf/openproject/blob/dev/.rubocop.yml) is included in the repository for your convenience. 
+
+Individual files can also be autocorrected with 
+
+```
+rubcocop -a ./path/to/file
+```
```
Comment (Azure7111) [inline CONTRIBUTING.md]: This will obviously not fix all of the stylecop offenses. We might want to specifically give the hint that it is useful to check manually nevertheless.

However, I'm not sure if that would be out of scope for this document and should be noted somewhere else.

Comment (floriank) [inline CONTRIBUTING.md]: Well, the goal imho should be to have it all in one place, otherwise we will have the Problem of multiple documents floating around having to be cared for as well.

We could rephrase that paragraph into a suggestion that this can be done to improve the style of one's own code.

Comment (Azure7111) [inline CONTRIBUTING.md]: At the end of the day my idea of abiding code style guidelines would simply be to not introduce new offenses (no matter what kind) and maybe even go the extra mile to remove some others as well. I usually like to remove the other offenses in the files I touched in a PR. However this could really put our rubocop configuration to the test, which might not be as desirable.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,7 +1,102 @@
...
+
+JavaScript style for the frontend can be verified via a `gulp` task in the `frontend` folder:
+
+```
+gulp lint
+```
+
+Additionally, we use HoundCI to verifiy the styles for Ruby, as well as for JavaScript. We will reject pull requests which do not meet the code style requirements.
```
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: Is reject the right word here?

We use reject for PRs without description, because in doubt we want to close them quickly... I don't think we would immediately close a PR for style violations, but ask the OP to fix the style issues first.

However, I could also misunderstand the matter and we do not immediately close any PRs.

If you think I am right I would propose something along those lines:

> We will not merge pull requests which...

Comment (floriank) [inline CONTRIBUTING.md]: This is correct. I'll change that wording.

File: CONTRIBUTING.md (hunk 10 above)
Comment (NobodysNightmare) [inline CONTRIBUTING.md]: I would not explicitly suggest that for the community (going the extra mile), but keep that as our mission...

This will not stop brave members of the community to fix the style nevertheless, but I think this is nothing we need to anticipate inour guidelines.

**Comment by floriank** [comment]
There is a pull request for the guidelines - https://github.com/floriank/openproject/pull/2

**Comment by floriank** [comment]
I discussed his suggestions with @marutosi and he has a point.

I would go along with him and open this matter for discussion:
- making a workflow with a forked repository mandatory for all members of the project
- creating pull requests to be worked on by multiple members 
- setting up refspec locally so that the PRs can be inspected locally
- removing the branch restriction from the `.travis.yml`

**Comment by floriank** [comment]
captainplanet4

Summoning @ulferts @kgalli @jonasheinrich @machisuji @linki for comments.

**Comment by machisuji** [comment]
I agree with @floriank and @marutosi on the matter of PRs through forks.
Capt. Planet would approve.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,7 +1,82 @@
...
+
+- Create a pull request against a branch of of the <opf/openproject> repository, containing a ***clear description*** of what the pull request attempts to change and/or fix. 
+
+If your pull request **does not contain a description** for what it does and what it's intentions are, we will reject it. If you are working on a specific work package from the [list](https://community.openproject.org/projects/openproject/work_packages?query_props=%7B%22c%22:%5B%22type%22,%22status%22,%22subject%22,%22assigned_to%22%5D,%22t%22:%22parent:desc%22,%22f%22:%5B%7B%22n%22:%22status_id%22,%22o%22:%22!%22,%22t%22:%22list_status%22,%22v%22:%5B%2217%22,%2223%22,%223%22,%2214%22,%226%22%5D%7D%5D,%22pa%22:1,%22pp%22:20%7D), you may include a link to that work package in the description, so we can track your work.
+
+We will then review your pull request. Please note that you can add commits after the pull request has been created by pushing to the branch in your fork.
+
+## Important notes
+
+To ensure a smooth workflow for, please take note of the following:
```
Comment (kgalli) [inline CONTRIBUTING.md]: I think the 'for' can be deleted. The rest sounds great so :+1:  I totally agree on PRs through forks.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,7 +1,82 @@
...
+
+### Issue tracking and coordination
+
+We use OpenProject for development coordination - please have a look at [the work packages list](https://community.openproject.org/projects/openproject/work_packages?query_props=%7B%22c%22:%5B%22type%22,%22status%22,%22subject%22,%22assigned_to%22%5D,%22t%22:%22parent:desc%22,%22f%22:%5B%7B%22n%22:%22status_id%22,%22o%22:%22!%22,%22t%22:%22list_status%22,%22v%22:%5B%2217%22,%2223%22,%223%22,%2214%22,%226%22%5D%7D%5D,%22pa%22:1,%22pp%22:20%7D) for upcoming features and reported bugs.
+
+### Etiquette and communication
+
+Lastly, be nice and respectful to each other. We are working hard to make OpenProject the best project management software there is and we are grateful for each contribution. 
+
+If you want to get in touch with us, there is also a [Gitter channel](https://gitter.im) to talk to us directly.
```
Comment (jonasheinrich) [inline CONTRIBUTING.md]: Should we specify the Gitter channel here?

File: CONTRIBUTING.md (hunk 12 above)
Comment (floriank) [inline CONTRIBUTING.md]: I think I had "for **everyone involved**" in mind.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,7 +1,82 @@
...
+
+## Important notes
+
+To ensure a smooth workflow for, please take note of the following:
+
+### Testing
+
+Please add tests to your code to verify functionality, especially if it is a new feature.
+
+Pull requests will be verified by TravisCI as well, but please run them locally as well. We have a lot of pull requests coming in and it takes some time to run the complete suite for each one.
```
Comment (jonasheinrich) [inline CONTRIBUTING.md]: Maybe we should say here that it would be best if the tests were green when run locally. So there is a chance that Travis needs to test this PR only once.

**Comment by myabc** [comment]
@floriank to reduce the amount of work Travis has to do, you can add `[ci skip]` to your commit messages.

**Comment by floriank** [comment]
@myabc I think I added that - but in a more demanding way:

> Please also use `[ci skip]` in your commit message to suppress builds which are not necessary (e.g. after fixing a typo in the README).

**Comment by myabc** [comment]
@floriank oh yes. / gitignore me

**Comment by floriank** [comment]
```
echo "@myabc" >> .gitignore
```

**Comment by Azure7111** [comment]
Should contributions to plugins follow the same guidelines?

**Comment by floriank** [comment]
Yeah, I guess so.

**Comment by kgalli** [comment]
I strongly recommend they do!

**Comment by floriank** [comment]
Question is - how would we redirect the users - just place a link to this `CONTRIBUTING.md` in the `CONTRIBUTING.md` in the plugin repos?

**Comment by floriank** [comment]
I will keep this PR open till Monday - after that it will be merged.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,7 +1,84 @@
...
+
+We want to keep the Pull request list as cleaned up as possible - we will aim close pull requests after an **inactivity period of 72 hours** (no comments, no further pushes) which are not labelled as `work in progress` by us.
+
+### Issue tracking and coordination
+
+We use OpenProject for development coordination - please have a look at [the work packages list](https://community.openproject.org/projects/openproject/work_packages?query_props=%7B%22c%22:%5B%22type%22,%22status%22,%22subject%22,%22assigned_to%22%5D,%22t%22:%22parent:desc%22,%22f%22:%5B%7B%22n%22:%22status_id%22,%22o%22:%22!%22,%22t%22:%22list_status%22,%22v%22:%5B%2217%22,%2223%22,%223%22,%2214%22,%226%22%5D%7D%5D,%22pa%22:1,%22pp%22:20%7D) for upcoming features and reported bugs.
+
+### Etiquette and communication
+
+Lastly, be nice and respectful to each other. We are working hard to make OpenProject the best project management software there is and we are grateful for each contribution.
```
Comment (myabc) [inline CONTRIBUTING.md]: @floriank Bundler now adds a Code of Conduct with its gem generator.
- https://github.com/bundler/bundler/blob/master/lib/bundler/templates/newgem/CODE_OF_CONDUCT.md.tt
- http://contributor-covenant.org/

Angular are also using it.

Might be something to consider.

Comment (floriank) [inline CONTRIBUTING.md]: Thanks, I'll have a look at it after workout :muscle:

**Comment by myabc** [comment]
:+1:

**Comment by myabc** [comment]
Two things that just occurred to me: 
- how (or will) Teatro instances work with this workflow? 
- ditto for Rubocop.

Sorry for late feedback.

**Comment by floriank** [comment]
Teatro and Hound should be available for all PRs - am I missing something? Rubocop can also be executed locally.

Features which would be developed by multiple developers can be created as a pull request set as `work in progress` - alternatively there is still the option to have a branch in the main repo, if desirable. the goal here should be to reduce the overall number.

**Comment by myabc** [comment]
> Teatro and Hound should be available for all PRs - am I missing something? Rubocop can also be executed locally.

Right, of course. I was missing something.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -1,7 +1,84 @@
...
+
+## Important notes
+
+To ensure a smooth workflow for everyone, please take note of the following:
+
+### Testing
+
+Please add tests to your code to verify functionality, especially if it is a new feature. 
+
+Pull requests will be verified by TravisCI as well, but please run them locally as well and make sure there are green before creating your pull request. We have a lot of pull requests coming in and it takes some time to run the complete suite for each one.
```
Comment (ulferts) [inline CONTRIBUTING.md]: typo:

`and make sure there are green`

->

```
and make sure they are green
```

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -1,7 +1,84 @@
...
+
+### Branching model
+
+The main development branch for upcoming releases is `dev`. For identifying the branch to create a pull request against, please refer to these rules:
+
+- If in doubt, create your pull request against `dev`. All new features, gem updates and bugfixes for the upcoming release should go into the `dev` branch.
+- Hotfixes should be created against the appropiate `release/*` branch. Backports for fixes also go against their specific `release/*` branch.
```
Comment (ulferts) [inline CONTRIBUTING.md]: The term "Hotfix" is not very specific. I think we should clarify it. For example: 

> Bugfixes for one of the actively supported versions of OpenProject should be issued against the respective branch. A fix for the current version (called "Hotfix" and the branch ideally being named `hotfix/XYZ`) should target `release/*` and a fix for the former version (called "Backport" and the branch ideally being named `backport/XYZ`) should target `backport/*`. We will try to merge hotfixes into dev branch but if that is no trivial task, we might ask you to create another PR for that.

Comment (floriank) [inline CONTRIBUTING.md]: :+1:

**Comment by ulferts** [comment]
I think it's good enough. There will be stuff we might want to improve but I rather have some guidelines now that we can improve later on. 

We can issue a PR changing and following the guidelines at the same time. WHOA.  

giphy

**Comment by floriank** [comment]
This is it folks. i am merging this. Thanks for all your contributions.

Here is a picture of my cat:

img_20150117_133204
